packages/protocol/contracts/verifiers/SP1Verifier.sol

```diff
@@ -0,0 +1,97 @@
+    ///      SP1VerifierGateway which can be used to verify proofs for any version of SP1.
+    ///      For the list of supported verifiers on each chain, see:
+    ///      https://github.com/succinctlabs/sp1-contracts/tree/main/contracts/deployments
+    address public verifier;
+
+    /// @notice The verification keys mappings for the proving programs.
+    mapping(bytes32 provingProgramVKey => bool trusted) public isProgramTrusted;
+
+    uint256[48] private __gap;
+
+    /// @dev Emitted when a trusted image is set / unset.
+    /// @param programVKey The id of the image
+    /// @param trusted The block's assigned prover.
+    event ProgramTrusted(bytes32 programVKey, bool trusted);
+
+    /// @dev Emitted when a new verifier address is set.
+    /// @param verifier The address of the verifier.
+    event NewVerifierAddress(address verifier);
+
+    error SP1_INVALID_PROGRAM_VKEY();
+
+    /// @notice Initializes the contract with the provided address manager.
+    /// @param _owner The address of the owner.
+    /// @param _addressManager The address of the AddressManager.
+    /// @param _verifier The address of the SP1Verifiers.
+    function init(
+        address _owner,
+        address _addressManager,
+        address _verifier
+    )
+        external
+        initializer
+    {
+        __Essential_init(_owner, _addressManager);
+
+        verifier = _verifier;
+    }
+
+    /// @notice Sets/unsets an the program's verification key as trusted entity
+    /// @param _programVKey The verification key of the program.
+    /// @param _trusted True if trusted, false otherwise.
+    function setProgramTrusted(bytes32 _programVKey, bool _trusted) external onlyOwner {
+        isProgramTrusted[_programVKey] = _trusted;
+
+        emit ProgramTrusted(_programVKey, _trusted);
+    }
+
+    /// @notice Sets the verifier contract.
+    /// @param _verifier The address of the verifier contract.
+    function setVerifierContract(address _verifier) external onlyOwner {
+        verifier = _verifier;
+
+        emit NewVerifierAddress(_verifier);
+    }
+
+    /// @inheritdoc IVerifier
+    function verifyProof(
+        Context calldata _ctx,
+        TaikoData.Transition calldata _tran,
+        TaikoData.TierProof calldata _proof
+    )
+        external
+        view
+    {
+        // Do not run proof verification to contest an existing proof
+        if (_ctx.isContesting) return;
+
+        // Decode will throw if not proper length/encoding
+        (bytes32 programVKey, bytes memory proof) = abi.decode(_proof.data, (bytes32, bytes));
+
+        if (!isProgramTrusted[programVKey]) {
+            revert SP1_INVALID_PROGRAM_VKEY();
+        }
+
+        uint64 chainId = ITaikoL1(resolve(LibStrings.B_TAIKO, false)).getConfig().chainId;
+
+        bytes memory encodedPublicInput = LibPublicInput.abiEncodePublicInputs(
+            _tran, address(this), address(0), _ctx.prover, _ctx.metaHash, chainId
+        );
```

Comment:
I think currently we actually do `sha256(keccak(public_inputs)`), so this would also just have to call `hashPublicInput`.

Reply:
And then convert bytes32 to bytes `abi.encode(hashedPublicInput)` ? Because `SPVerifier` expects `bytes`. But then we are where we are now :) So i'm not sure i follow your comment.

Reply:
Done the "regular" way.